solc/CommandLineInterface.cpp

```diff
@@ -187,7 +191,11 @@ void CommandLineInterface::handleBinary(string const& _contract)
 
 void CommandLineInterface::handleOpcode(string const& _contract)
 {
-	solAssert(m_options.input.mode == InputMode::Compiler || m_options.input.mode == InputMode::CompilerWithASTImport, "");
+	solAssert(
+		m_options.input.mode == InputMode::Compiler ||
+		m_options.input.mode == InputMode::CompilerWithASTImport ||
+		m_options.input.mode == InputMode::CompilerWithEvmAssemblyJsonImport, ""
+	);
```

Comment:
Maybe you could create a constant representing all three modes? This used to be just a quick assert and now it got pretty verbose in quite a few places.
```suggestion
	solAssert(compilationInputModes.count(m_options.input.mode) > 0);
```

Reply:
I would prefer to be more explicit. But I agree that it became large and too verbose. Your suggestion is not really reflecting what modes are exactly allowed. Maybe I will create some helper functions, so that e.g. something like this is possible `solAssert(or(m_options.input.mode, {InputMode::Compiler, InputMode::CompilerWithASTImport, InputMode::CompilerWithEvmAssemblyJsonImport})` - where `{InputMode::Compiler, InputMode::CompilerWithASTImport, InputMode::CompilerWithEvmAssemblyJsonImport}` could be defined somewhere and just get reused. Let me think a bit more to find maybe a even better solution. But you're totally right, it's getting too verbose.

Reply:
Ok, I introduced `any_of` and `none_of` that are now used in these macros. See commit https://github.com/ethereum/solidity/pull/12834/commits/d6f7aeac41bf4db2b56cfcffdee160a7477fa60a. I added those function in `solc/CommandLineParser.h`. Maybe we should move those to a more generic place. What would you say, where is a good place for these functions?

Reply:
> Your suggestion is not really reflecting what modes are exactly allowed.

`compilationInputModes` was meant to be a constant `map` or `set` (I was hoping it was obvious from `count()`) that has all the `Compiler*` modes set. With that it would work, wouldn't it?

And sure, `any_off`/`none_of` are fine but you would not need them if you used a `set` :)

> What would you say, where is a good place for these functions?

We already have [`std::ranges::all_of`, `std::ranges::any_of`, `std::ranges::none_of`](https://en.cppreference.com/w/cpp/algorithm/ranges/all_any_none_of) in ranges v3. Why not use those?

Reply:
True! That should definitely work!

Reply:
I just looked now superficially into this. But `std::ranges::all_of` does not work on single elements, it need to be a "collection". The usage of `ranges` would look like `solAssert(ranges::any_of(std::set<InputMode>{m_options.input.mode}, IsValidInputMode));` where `IsValidInputMode` is a predicate instead of a vector or set. Not sure whether this is nice. However, I could use `std::ranges::any_of` to implement `any_of`. 

However, I think I finally really understood your `compilationInputModes.count(m_options.input.mode)` suggestion now. I totally misread it. I agree, that's a good solution. So I guess I will go with your suggestion and remove the `any_of` stuff totally.

Reply:
Ok, I removed the `any_of` stuff - I used your suggestion now.